core/module.go

```diff
@@ -88,6 +119,190 @@ func (mod *Module) Dependencies() []Mod {
+	// in the dependencies arg here
+	depSet := make(map[string]*ModuleDependency)
+	for _, dep := range mod.DependencyConfig {
+		symbolic, err := dep.Source.Self.Symbolic()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get symbolic source ref: %w", err)
+		}
+		depSet[symbolic] = dep
+	}
+	for _, newDep := range clonedDependencies {
+		symbolic, err := newDep.Source.Self.Symbolic()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get symbolic source ref: %w", err)
+		}
+		depSet[symbolic] = newDep
+	}
+
+	mod.DependencyConfig = make([]*ModuleDependency, 0, len(depSet))
+	for _, dep := range depSet {
+		mod.DependencyConfig = append(mod.DependencyConfig, dep)
+	}
+
+	refStrs := make([]string, 0, len(mod.DependencyConfig))
+	for _, dep := range mod.DependencyConfig {
+		refStr, err := dep.Source.Self.RefString()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get ref string for dependency: %w", err)
+		}
+		refStrs = append(refStrs, refStr)
+	}
+	sort.Slice(mod.DependencyConfig, func(i, j int) bool {
+		return refStrs[i] < refStrs[j]
+	})
+
+	mod.DependenciesField = make([]dagql.Instance[*Module], len(mod.DependencyConfig))
+	eg = errgroup.Group{}
+	for i, depCfg := range mod.DependencyConfig {
+		i, depCfg := i, depCfg
+		eg.Go(func() error {
+			err := srv.Select(ctx, depCfg.Source, &mod.DependenciesField[i],
+				dagql.Selector{
+					Field: "asModule",
+				},
+				dagql.Selector{
+					Field: "withName",
+					Args: []dagql.NamedInput{
+						{Name: "name", Value: dagql.NewString(depCfg.Name)},
+					},
+				},
+				dagql.Selector{
+					Field: "initialize",
+				},
+			)
+			if err != nil {
+				return fmt.Errorf("failed to initialize dependency module: %w", err)
+			}
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		if errors.Is(err, dagql.ErrCacheMapRecursiveCall) {
+			err = fmt.Errorf("module %s has a circular dependency: %w", mod.NameField, err)
+		}
+		return nil, fmt.Errorf("failed to load updated dependencies: %w", err)
+	}
+	// fill in any missing names if needed
+	for i, dep := range mod.DependencyConfig {
+		if dep.Name == "" {
+			dep.Name = mod.DependenciesField[i].Self.Name()
+		}
+	}
+
+	// Check for loops by walking the DAG of deps, seeing if we ever encounter this mod.
+	// Modules are identified here by their *Source's ID*, which is relatively stable and
+	// unique to the module.
+	// Note that this is ultimately best effort, subtle differences in IDs that don't actually
+	// impact the final module will still result in us considering the two modules different.
+	// E.g. if you initialize a module from a source directory that had a no-op file change
+	// made to it, the ID used here will still change.
+	selfID := mod.Source.ID().String()
+	memo := make(map[string]struct{}) // set of id strings
+	var visit func(dagql.Instance[*Module]) error
+	visit = func(inst dagql.Instance[*Module]) error {
+		instID := inst.Self.Source.ID().String()
```

Comment:
nit: using `.Digest` here would be more consistent with the general practice, and is approximately the same thing aside from the pesky `error` return value (which `String` just drops internally)